personalrobotics/ada_feeding
# Finalize learning capabilities (not including posthoc)

# Description

This PR merges learning capabilities into the feeding tree. About 90% of the learning code was implemented previously, but some functionality tweaks needed to be addressed, and some bugs needed fixing. Specifically:

1. All learning algorithms are now usable by adding a flag to start.py
2. The code now reports what action is chosen at each selection step
3. The bug where two different actions are chosen when breaking the pi-symmetry has been addressed; now, an action is chosen on the first `move_above_plan`, and that same action is used on the second. This addresses this issue: https://github.com/personalrobotics/ada_feeding/issues/191. 
4. As a note, only the "no context" and "SPANet" context work; the "color" context (color averaging) does not. The code is currently set to use SPANet context; you can set to "no context" in https://github.com/personalrobotics/ada_feeding/blob/ros2-devel/ada_feeding_action_select/config/policies.yaml, but this effectively negates learning as the robot has no visual context to learn from. The posthoc ft context is also not implemented.

# Testing procedure

To test this code:

1. Run the start.py code with each learning algorithm name. E.g. if running on the robot, run `python3 src/ada_feeding/start.py --policy=linucb_noposthoc`. The learning algorithm options (without posthoc context or color context) are: ['constant', 'random_noposthoc', 'greedy_noposthoc', 'egreedy_noposthoc', 'linucb_noposthoc'].  For learning, we suggest running "linucb_noposthoc". For a constant single action, use "constant" (the default if no argument is passed in). To use only the best actions, use "greedy_noposthoc". The code will load the latest checkpoint in the data file (this locations is already set/saved on Lovelace). 
2. Check the action output, printed to the screen as `self.logger.info(f"Chosen action index: {index}")` from line 150 in https://github.com/personalrobotics/ada_feeding/blob/taylorkf/learning-inclusion/ada_feeding/ada_feeding/behaviors/acquisition/compute_action_constraints.py. Confirm that this is only printed to the screen once per bite; if printed twice, this implies that the code to break pi-symmetry is selecting an action twice, rather than re-using the first chosen action.
3. Ensure that checkpoints are being saved to the data/ folder on lovelace.


# Before opening a pull request

# Before Merging
